Tensile/Source/lib/source/ContractionSolution.cpp

```diff
@@ -40,6 +41,119 @@
 
 namespace Tensile
 {
+    namespace streamk
+    {
+        namespace math
+        {
+            /**
+             * Performs `(n + d - 1) / d` (unsafe).
+             */
+            template <typename N, typename D>
+            __device__ __host__ inline constexpr N ceil_div(N n, D d)
+            {
+                // Static cast to undo integral promotion.
+                return static_cast<N>((n + d - 1) / d);
+            }
+
+            /**
+             * Performs `(n + d - 1) / d`, but is robust against the case where
+             * `(n + d - 1)` would overflow.
+             */
+            template <typename N, typename D>
+            __device__ __host__ inline constexpr N safe_ceil_div(N n, D d)
+            {
+                // Static cast to undo integral promotion.
+                return static_cast<N>(n / d + (n % d != 0 ? 1 : 0));
+            }
+        }  // namespace math
+
+        constexpr int num_iters_per_cta(int BLK_M, int BLK_N, int BLK_K, int m, int n, int k, int g)
+        {
+            return math::ceil_div(math::ceil_div(m, BLK_M) * math::ceil_div(n, BLK_N) * math::ceil_div(k, BLK_K), g);
+        }
+
+        constexpr int number_of_output_tiles(int BLK_M, int BLK_N, int m, int n)
+        {
+            int m_tiles =  math::ceil_div(m, BLK_M);
+            int n_tiles = math::ceil_div(n, BLK_N);
+            return m_tiles * n_tiles;
+        }
+
+        constexpr int num_fixup_peers(int BLK_K, int k, int g, int iters_per_cta)
+        {
+            return math::ceil_div(math::ceil_div(k, BLK_K), iters_per_cta);
+        }
+
+        std::tuple<double,int,int>
+        predicted_runtime(int BLK_M, int BLK_N, int BLK_K, int m, int n, int k, int g, int a, int b, int c, int d)
+        {
+            int iters_per_cta = num_iters_per_cta(BLK_M, BLK_N, BLK_K, m, n, k, g);
+            int fixup_peers = num_fixup_peers(BLK_K, k, g, iters_per_cta);
+
+            return {a + (b * (fixup_peers > 1)) + (c * iters_per_cta) + (d * (fixup_peers - 1)), iters_per_cta, fixup_peers};
+        }
+
+        int best_predicted_grid_size(int BLK_M,
+                                     int BLK_N,
+                                     int BLK_K,
+                                     int m,
+                                     int n,
+                                     int k,
+                                     int grid_start = 1,
+                                     int grid_end = 304)
+        {
+            static const bool debug = Debug::Instance().printStreamKGridInfo();
+            // Timing in us @ 1100/900 MHz
+            double a = 5.04 + 8.3; // 5544 + 9136;  // 5.04 us
```

Comment:
I do not understand what these values mean, but I am feeling this logic is too specialized to specific arch.